ext/mbstring/mbstring.c

```diff
@@ -2183,6 +2190,202 @@ PHP_FUNCTION(mb_output_handler)
 }
 /* }}} */
 
+/* {{{ proto array mb_str_split(string str [, int split_length] [, string encoding])
+ Convert a multibyte string to an array. If split_length is specified,
+ break the string down into chunks each split_length characters long. */
+
+/* structure to pass split params to the callback */
+struct mbfl_split_params {
+    zval *return_value; /* php function return value structure pointer */
+    mbfl_string *result_string; /* string to store result chunk */
+    size_t mb_chunk_length; /* actual chunk length in chars */
+    size_t split_length; /* split length in chars */
+    mbfl_convert_filter *next_filter; /* widechar to encoding converter */
+};
+
+/* callback function to fill split array */
+static int mbfl_split_output(int c, void *data)
+{
+    struct mbfl_split_params *params = (struct mbfl_split_params *)data; /* cast passed data */
+
+    (*params->next_filter->filter_function)(c, params->next_filter); /* decoder filter */
+
+    if(params->split_length == ++params->mb_chunk_length) { /* if current chunk size reached defined chunk size or last char reached */
+        mbfl_convert_filter_flush(params->next_filter);/* concatenate separate decoded chars to the solid string */
+        mbfl_memory_device *device = (mbfl_memory_device *)params->next_filter->data; /* chars container */
+        mbfl_string *chunk = params->result_string;
+        mbfl_memory_device_result(device, chunk); /* make chunk */
+        add_next_index_stringl(params->return_value, chunk->val, chunk->len); /* add chunk to the array */
+        efree(chunk->val);
+        params->mb_chunk_length = 0; /* reset mb_chunk size */
+    }
+    return 0;
+}
+
+PHP_FUNCTION(mb_str_split)
+{
+    zend_string *str, *encoding = NULL;
+    size_t mb_len, chunks, chunk_len;
+    char unsigned const *last, *p;
+    mbfl_string string, result_string;
+    const mbfl_encoding *mbfl_encoding;
+    zend_long split_length = 1;
+
+    ZEND_PARSE_PARAMETERS_START(1, 3)
+            Z_PARAM_STR(str)
+            Z_PARAM_OPTIONAL
+            Z_PARAM_LONG(split_length)
+            Z_PARAM_STR(encoding)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (split_length <= 0) {
+        php_error_docref(NULL, E_WARNING, "The length of each segment must be greater than zero");
+        RETURN_FALSE;
+    }
+
+    /* fill mbfl_string structure */
+    string.no_language = MBSTRG(language);
+    string.encoding = php_mb_get_encoding(encoding);
+    if (!string.encoding) {
+        RETURN_FALSE;
+    }
+
+    p = (char unsigned *)ZSTR_VAL(str); /* string cursor pointer */
+    string.val = (char unsigned *)p;
+    string.len = ZSTR_LEN(str);
+    last = string.val + string.len; /* last string char pointer */
+
+    /* check if split_length > string.len */
+    if(split_length > string.len){
```

Comment:
This check does not really fix the issue. E.g. if you have a string of length 10000001 characters and a split length of 10000000, the problem will be back. It's really necessary to check this on each iteration.

Reply:
Got it.